spatialconnect/src/main/java/com/boundlessgeo/spatialconnect/dataAdapter/GeoPackageAdapter.java

```diff
@@ -47,114 +81,552 @@
+                                    }
+                                    catch (IOException e) {
+                                        subscriber.onError(e);
+                                    }
+                                    finally {
+                                        response.body().close();
+                                    }
+                                }
+                            });
+                }
+            }
+        });
+    }
+
+
+    /**
+     * For each query, we use a <a href="https://en.wikipedia.org/wiki/Correlated_subquery">correlated sub query</a>
+     * on the R*-tree index to optimize the query.
+     *
+     * @param queryFilter
+     * @return
+     */
+    public Observable<SCSpatialFeature> query(final SCQueryFilter queryFilter) {
+        final List<SCGpkgFeatureSource> sources = getFeatureSources();
+        return Observable.create(new Observable.OnSubscribe<SCSpatialFeature>() {
+            @Override
+            public void call(final Subscriber<? super SCSpatialFeature> subscriber) {
+                // if there are no layer names supplied in the query filter, then search on all tables
+                List<String> layerNames = queryFilter.getLayerIds();
+                if (layerNames.size() == 0) {
+                    layerNames = getFeatureSourceNames();
+                }
+                Observable.from(layerNames)
+                        .filter(new Func1<String, Boolean>() {
+                            @Override
+                            public Boolean call(String layer) {
+                                return sources.contains(getFeatureSourceByName(layer));
+                            }
+                        })
+                        .flatMap(new Func1<String, Observable<SCSpatialFeature>>() {
+                            @Override
+                            public Observable<SCSpatialFeature> call(final String layerName) {
+                                final SCGpkgFeatureSource source = getFeatureSourceByName(layerName);
+                                return gpkg.db.createQuery(
+                                        layerName,
+                                        String.format(
+                                                "SELECT %s FROM %s WHERE %s IN (%s) LIMIT %d",
+                                                getSelectColumnsString(source),
+                                                layerName,
+                                                source.getPrimaryKeyName(),
+                                                createRtreeSubQuery(source, queryFilter.getPredicate().getBoundingBox()),
+                                                100
+                                        )
+                                ).flatMap(getFeatureMapper(source));
+                            }
+                        })
+                        .subscribe(new Action1<SCSpatialFeature>() {
```

Comment:
This is superfluous..change to just return the observable (instead of creating a new one)